core/src/main/java/com/linecorp/armeria/internal/common/AbstractKeepAliveHandler.java

```diff
@@ -120,14 +127,22 @@ protected AbstractKeepAliveHandler(Channel channel, String name, Timer keepAlive
     public final void initialize(ChannelHandlerContext ctx) {
         // Avoid the case where destroy() is called before scheduling timeouts.
         // See: https://github.com/netty/netty/issues/143
-        if (isInitialized) {
+        if (keepAliveState != KeepAliveState.NONE) {
             return;
         }
-        isInitialized = true;
+
+        keepAliveState = KeepAliveState.INITIALIZED;
+        if (connectionEventListener != null) {
+            connectionEventListener.connectionOpened();
```

Comment:
It seems like if `initialize` is called due to `channelActive` or `handlerAdded`, the `HttpSession#protocol` may not exist. Can you check this point?

Reply:
You are right. connectionEventListener.connectionOpened() should be called after sessionPromise is completed.
Furthermore, (to support maxUnfinishedRequest filter) this should called before checking unfinished request.
And I was struggling to solve this problem. 🫠

```java
final class HttpChannelPool implements AsyncCloseable {
....
    private void notifyConnect(SessionProtocol desiredProtocol,
                               PoolKey key, Future<Channel> future,
                               ChannelAcquisitionFuture promise,
                               ClientConnectionTimingsBuilder timingsBuilder) {
              ....
              // In the current implementation(master branch), connectionOpen is called at here

               if (session.incrementNumUnfinishedResponses()) {
                    if (protocol.isMultiplex()) {
                        final Http2PooledChannel pooledChannel = new Http2PooledChannel(channel, protocol);
                        addToPool(protocol, key, pooledChannel);
                        promise.complete(pooledChannel);
                    } else {
                        promise.complete(new Http1PooledChannel(channel, protocol, key));
                    }
                } else {
                    // Server set MAX_CONCURRENT_STREAMS to 0, which means we can't send anything.
                    channel.close();
                    promise.completeExceptionally(
                            UnprocessedRequestException.of(RefusedStreamException.get()));
                }
              ...
        }
```